sql: panic in render expr during grouping

I am running a 3 node cockroach secure cluster.

The table definition is
`root@sl73commapd003.visa.com:9080> show columns from api;
+-----------+-----------+-------+----------------+
|   Field   |   Type    | Null  |    Default     |
+-----------+-----------+-------+----------------+
| id        | INT       | false | unique_rowid() |
| date      | TIMESTAMP | true  | NULL           |
| caller    | STRING    | true  | NULL           |
| component | STRING    | true  | NULL           |
| method    | STRING    | true  | NULL           |
| url       | STRING    | true  | NULL           |
| count     | INT       | true  | NULL           |
| min       | INT       | true  | NULL           |
| median    | INT       | true  | NULL           |
| perc90    | INT       | true  | NULL           |
| perc95    | INT       | true  | NULL           |
| perc96    | INT       | true  | NULL           |
| perc97    | INT       | true  | NULL           |
| perc98    | INT       | true  | NULL           |
| perc99    | INT       | true  | NULL           |
| max       | INT       | true  | NULL           |
| status    | INT       | true  | NULL           |
+-----------+-----------+-------+----------------+`


The SQL query that cause the node to terminate:

`select avg(median),avg(perc98),avg(max) from api where url like '%/rsrv_eag/1/0/devices/{ID}/AccountVerificationRequest%' group by 1;`

The error log file is attached
[error.log.zip](https://github.com/cockroachdb/cockroach/files/501618/error.log.zip)

Panic message:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x974e61]

goroutine 510 [running]:
panic(0x159d840, 0xc4200120a0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Recover(0xc4203bf9e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:176 +0x6e
panic(0x159d840, 0xc4200120a0)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/sql/parser.(*FuncExpr).Eval(0xc42032c5a0, 0xc420b338d8, 0x20a2240, 0xc420dc6dd8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:1833 +0x2d1
github.com/cockroachdb/cockroach/sql.(*selectNode).renderRow(0xc420cf7560, 0x0, 0xc420b338d8)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:574 +0x78
github.com/cockroachdb/cockroach/sql.(*selectNode).Next(0xc420cf7560, 0xc42001ae80, 0x2098a00, 0xc42001ae80)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:156 +0x155
github.com/cockroachdb/cockroach/sql.(*groupNode).Next(0xc420613760, 0xc420613760, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/group.go:273 +0x475
github.com/cockroachdb/cockroach/sql.(*selectTopNode).Next(0xc42001af00, 0xc42018ac00, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/select_top.go:162 +0x34
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc4200ba0f0, 0x2099480, 0xc4206787b0, 0xc420b338b8, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:1131 +0x299
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc4200ba0f0, 0x2099480, 0xc4206787b0, 0xc420b338b8, 0x101, 0xc420b33828, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:986 +0x261
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc4200ba0f0, 0xc42027e880, 0x1, 0x1, 0xc420b338b8, 0xc420b33828, 0x610101, 0x4, 0x0, 0xc4207a0f70, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:734 +0x2fb
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc4200ba0f0, 0xc420b338b8, 0x1, 0xc420b33828, 0xc4203a9a60, 0xc42027e880, 0x1, 0x1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:663 +0xeb
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func2(0xc4205b5a20, 0xc4203a9a60, 0x4000000000000000, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:525 +0x213
github.com/cockroachdb/cockroach/internal/client.(*Txn).Exec(0xc4205b5a20, 0xc420290101, 0x1478e3cf7f472168, 0x0, 0xc420299080, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:504 +0x23d
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc4200ba0f0, 0xc420b33800, 0xc4206e7000, 0x85, 0x0, 0x0, 0x0, 0x2084c00)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:539 +0x421
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc4200ba0f0, 0xc420b33800, 0xc4206e7000, 0x85, 0x0, 0xeac659, 0xc4206e7000, 0x86, 0x1000)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:400 +0xab
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc4200fd380, 0x7f832a7ef8f0, 0xc420db3140, 0xc4206e7000, 0x85, 0x0, 0x0, 0x0, 0x0, 0x6fb901, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:641 +0x7c
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc4200fd380, 0x7f832a7ef8f0, 0xc420db3140, 0xc4200fd3a8, 0x8a, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:320 +0xb3
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc4200fd380, 0xc4203a94a0, 0xc4203a94a0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:275 +0x85c
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc4201ba2c0, 0x209f1a0, 0xc42077a000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:240 +0x5ed
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x209f200, 0xc420c69b80)
	/go/src/github.com/cockroachdb/cockroach/server/server.go:418 +0x4e
github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1(0xc4203bf9e0, 0xc420028450, 0x209f200, 0xc420c69b80, 0xc4203c6520)
	/go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:138 +0x95
created by github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith
	/go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:140 +0x277

```

--- a/sql/group.go
+++ b/sql/group.go
@@ -45,9 +45,11 @@ func (p *planner) groupBy(n *parser.SelectClause, s *selectNode) (*groupNode, er
 	// that determination is made during validation, which will require matching
 	// expressions.
 	for i := range groupBy {
-		if err := p.parser.AssertNoAggregationOrWindowing(groupBy[i], "GROUP BY"); err != nil {
-			return nil, err
-		}
+		// Hold on to the original raw GROUP BY expression, which will be replaced by the
+		// original raw SELECT expression in the case of a GROUP BY <ordinal>. We will
+		// perform some verification on this which requires that the expression has not
+		// been modified by upper layers.
+		rawExpr := groupBy[i]
 
 		// We do not need to fully analyze the GROUP BY expression here
 		// (as per analyzeExpr) because this is taken care of by addRender
@@ -65,8 +67,13 @@ func (p *planner) groupBy(n *parser.SelectClause, s *selectNode) (*groupNode, er
 			return nil, err
 		} else if col >= 0 {
 			groupBy[i] = s.render[col]
+			rawExpr = n.Exprs[col].Expr
 		} else {
 			groupBy[i] = resolved
+		}
+
+		if err := p.parser.AssertNoAggregationOrWindowing(rawExpr, "GROUP BY"); err != nil {
+			return nil, err
 		}
 	}
 

--- a/sql/testdata/aggregate
+++ b/sql/testdata/aggregate
@@ -82,6 +82,9 @@ SELECT COUNT(*), k FROM kv GROUP BY 2
 query error aggregate functions are not allowed in GROUP BY
 SELECT * FROM kv GROUP BY v, COUNT(DISTINCT w)
 
+query error aggregate functions are not allowed in GROUP BY
+SELECT COUNT(DISTINCT w) FROM kv GROUP BY 1
+
 query error aggregate functions are not allowed in RETURNING
 INSERT INTO kv (k, v) VALUES (99, 100) RETURNING sum(v)
 

--- a/sql/testdata/window
+++ b/sql/testdata/window
@@ -22,6 +22,9 @@ INSERT INTO kv VALUES
 query error window functions are not allowed in GROUP BY
 SELECT * FROM kv GROUP BY v, COUNT(w) OVER ()
 
+query error window functions are not allowed in GROUP BY
+SELECT COUNT(w) OVER () FROM kv GROUP BY 1
+
 query error window functions are not allowed in RETURNING
 INSERT INTO kv (k, v) VALUES (99, 100) RETURNING sum(v) OVER ()
 

